provider/core.rb

```diff
@@ -70,14 +62,28 @@ class FileTemplate
     attr_accessor :async
     attr_accessor :resource_name
 
-    def initialize(options)
-      options.each { |k, v| method("#{k}=").call(v) }
+    class << self
+      # Construct a new FileTemplate based on a resource object
+      def file_for_resource(output_folder, object, version, config, env)
```

Comment:
I'm a little unclear how this is different than just calling the constructor. Doesn't the constructor handle setting `object` and `config`?

Reply:
It's delegating to a shared constructor between this and the `compile_file_list` method. We have two types of FileTemplates that share some fields, so I changed the constructor to set just the shared fields. Now the `compile_file_list` can call the constructor directly with its arguments, and compiling a resource file can call this method which clearly shows what fields are different between the two.

I prefer this more structured approach over passing a hash and setting fields based on the contents of the hash, but I'm happy to discuss it